packages/imperative/src/plugins/PluginManagementFacility.ts

```diff
@@ -257,89 +258,87 @@ export class PluginManagementFacility {
+            overrideDispNm = settingVal as string;
+            overridePluginNm = settingVal as string;
+
+            if (settingVal !== false && settingVal !== ImperativeConfig.instance.hostPackageName) {
                 Logger.getImperativeLogger().debug(
-                    `PluginOverride: Attempting to overwrite "${setting}" with value provided by plugin "${pluginName}"`
+                    `Attempting to replace "${settingNm}" with an override named "${overridePluginNm}"`
                 );
-                if (!Object.prototype.hasOwnProperty.call(loadedOverrides, pluginName)) {
-                    // the plugin name specified in our settings is not available
-                    const overrideErrMsg = `You attempted to override the "${setting}" setting ` +
-                        `with a plugin named "${pluginName}" that is not installed and loadable.` +
-                        `\nWe will use a "${setting}" that purposely fails until you reconfigure.`;
-                    Logger.getImperativeLogger().error(overrideErrMsg);
-
-                    /* We need to assign a class (which always shows an error when
-                     * the CLI tries to use credentials) into the current override setting.
-                     * We also need to embed our error message into that class. We cannot
-                     * create a new object from a class and pass the error into its
-                     * constructor, because the CredentialManagerFactory takes a class and
-                     * it calls the constructor of our supplied class. Thus we need an
-                     * anonymous class so that we can access our 'overrideErrMsg' variable.
-                     * Our trick is that we simply throw an error in the constructor
-                     * of our anonymous class. The CredentialManagerFactory catches
-                     * our error, and places it into its InvalidCredentialManager,
-                     * which in turn shows our error every time the CLI tries to use
-                     * credentials. Finally since lint complains about more than one
-                     * class in a file, we have to temporarily turn off that lint error.
+
+                if (settingNm === CredentialManagerOverride.CRED_MGR_SETTING_NAME) {
+                    /* For credMgr override, the value of the setting is the override display name.
+                     * We must use the display name to find this override within our loadedOverrides.
+                     * We must find the plugin name from within our known credMgr overrides.
                      */
-                    (this.mPluginOverrides as any)[setting] = class {
-                        constructor() {
-                            throw overrideErrMsg;
-                        }
-                    };
+                    const credMgrInfo = CredentialManagerOverride.getCredMgrInfoByDisplayName(overrideDispNm);
```

Comment:
The following guard encloses the code that you identified. The value of ImperativeConfig.instance.hostPackageName is "@zowe/cli", so we never get to the code that you identified.
 ```
if (settingVal !== false && settingVal !== ImperativeConfig.instance.hostPackageName) {
```
However, when the settingVal is "Zowe CLI", it does reach the code that you identified. That code passess successfully because "Zowe CLI" is in our list of known credMgrs.
 
An even bigger problem occurs with "Zowe CLI" when we reach this logic a little later in the same block of code.

```
if (!Object.prototype.hasOwnProperty.call(loadedOverrides, overridePluginNm)) {
    // The overrideName specified in our settings is not available from any plugin.
    this.useOverrideThatFails(settingNm, overrideDispNm, overridePluginNm,
        `No plugin has been installed that overrides '${settingNm}' with '${overrideDispNm}'.` +
        "\nPlugins that provide overrides are:\n" + JSON.stringify(loadedOverrides, null, 2)
    );
    continue;
}
```

The "Zowe CLI" setting name has no plugin associatated with it (because it is built-in), so we get this error with our new naming. Our automated tests did not fail, but I easily caused the error this morning with a manual test.

I think it may be possible to place both "@zowe/cli" and "Zowe CLI" into our list of known credMgrs and include more checks for known built-in names in our logic. I feel that such an approach is becoming more confusing to explain and more prone to an errors in our logic. I think we will provide a more stable product if we just revert back to only recognizing the original name of "@zowe/cli". I will revert the code back to its original behavior. I also plan to change the first if-statement above from checking ImperativeConfig.instance.hostPackageName to checking for CredentialManagerOverride.DEFAULT_CRED_MGR_NAME.

Reply:
I agree that some-other-cli will not be successfully loaded as a default credential manager. However, Zowe Product Management decided some time ago that the Zowe client will only support plugins, not additional CLIs. As such, we should not break anything that currently exists. The idea that the imperative library can be used generically by any CLI has slowly morphed into really only being used for Zowe CLI. Some proposals have even been floated to include the imperative repository into the Zowe repository. Obviously that last action would have to occur in some future, major, breaking release. My position is that our ability to handle both plugins and ZE extensions by confirming that the supplied name has a related plugin or extension outweighs the limitation that the default credential manager must be supplied by Zowe CLI.

We could add an extra test to confirm that the configured name is either the default name (@zowe/cli), or the configured name is identical to the current CLI's package name. Do you feel that such a change is justified?

Reply:
The way that the current CLI's package name is handled in this PR seems sound 👍 

My concern is related to extenders who are developing a new credential manager plugin. While I agree we don't need to support them in the `recordCredMgrInConfig` method, I think the `PluginManagementFacility` should allow anyone to test their plugin in Zowe CLI. When I test the changes in this PR with a sample credential manager plugin, it doesn't work.

Steps to reproduce the issue:
* Create a new CLI plugin that defines a CredentialManager override and give it a name like `@zowe/sample-credential-manager-for-zowe-cli`.
* Build the sample plugin and install it into Zowe CLI.
* Manually update CredentialManager property in `~/.zowe/settings/imperative.json` file to match the plugin name (`@zowe/sample-credential-manager-for-zowe-cli`).
* Run a CLI command and get the following error:
    ```
    [2023/03/23 16:38:16.394] [ERROR] [CredentialManagerFactory.js:149] Failed to load the credential manager named "@zowe/sample-credential-manager-for-zowe-cli"
    [2023/03/23 16:38:16.415] [ERROR] [CredentialManagerFactory.js:151] Unable to override "CredentialManager" with "@zowe/sample-credential-manager-for-zowe-cli" from plugin "@zowe/sample-credential-manager-for-zowe-cli"
    Reason = The "CredentialManager" named "@zowe/sample-credential-manager-for-zowe-cli" is not a known credential manager override.
    We will use a "CredentialManager" that purposely fails until you reconfigure.
    ```
* If you run the same CLI command using Zowe CLI v7.11.0, it successfully loads the sample plugin and uses its credential manager.

Sorry for any confusion I've caused around this topic, perhaps we can discuss more tomorrow in person? 🙂

Reply:
Now I understand. You were coming at it from a different angle. Andrew expressed similar concerns, but I overlooked this particular if-condition. We may have boxed ourselves in somewhat (from an early testing perspective) with our design plan to keep a known list. We can brainstorm whether we can do anything that can help testing.

Reply:
Perhaps we could we implement something like the following pseudo code:
```
If override setting value is a display name defined in our map (e.g., `Secrets for Kubernetes`) then use it
Else if override setting value is a plugin name defined in our map (e.g., `@zowe/secrets-for-kubernetes-for-zowe-cli`) then use it
Else if override setting value is a plugin name installed in the CLI (e.g., `@zowe/sample-credential-manager-for-zowe-cli`) then issue a warning that the credential manager is unknown but still use it
Else use the invalid credential manager
```